### PR TITLE
[Refactor] Support CID

### PR DESCRIPTION
**Pull request by Ben-Louis**
## Modification

1. add CIDHead
2. add a codec for CID named 'DecoupledHeatmap', which means that the heatmaps for each instance are decoupled and represented on different channels
3. add configs

TODO:
- [x] add unittests 
- [x] add docstrings

## Checklist

**Before PR**:

- [ ] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmpose/blob/master/.github/CONTRIBUTING.md) to create this PR.
- [ ] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmpose/blob/master/.github/CONTRIBUTING.md) are used to fix the potential lint issues.
- [ ] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] New functionalities are covered by complete unit tests. If not, please add more unit tests to ensure correctness.
- [ ] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmpose/models/heads/heatmap_heads/cid_head.py [hunk 1]
```diff
@@ -0,0 +1,727 @@
...
+                instance_heatmaps_flip = \
+                    instance_heatmaps_flip[:, flip_indices, :, :]
+                instance_heatmaps = (instance_heatmaps +
+                                     instance_heatmaps_flip) / 2.0
+            instance_heatmaps = smooth_heatmaps(
+                instance_heatmaps, test_cfg.get('blur_kernel_size', 3))
+
+            preds = self.decode((instance_heatmaps, instance_scores[:, None]))
+            preds = InstanceData.cat(preds)
+            preds.keypoints[..., :2] += 1.5
```
Comment (jin-s13) [inline mmpose/models/heads/heatmap_heads/cid_head.py:581]: why 1.5?

File: mmpose/models/heads/heatmap_heads/cid_head.py [hunk 2]
```diff
@@ -0,0 +1,727 @@
...
+
+        self.channel_attention = ChannelAttention(in_channels, gfd_channels)
+        self.spatial_attention = SpatialAttention(in_channels, gfd_channels)
+        self.fuse_attention = build_conv_layer(
+            dict(
+                type='Conv2d',
+                in_channels=gfd_channels * 2,
+                out_channels=gfd_channels,
+                kernel_size=1))
+        self.heatmap_conv = build_conv_layer(
```
Comment (jin-s13) [inline mmpose/models/heads/heatmap_heads/cid_head.py:334]: In mmpose 0.x, heatmap_conv.bias and keypoint_center_conv.bias is initialized as bias_value. Do we still have this feature in 1.x?

bias_value = -math.log((1 - self.prior_prob) / self.prior_prob)
torch.nn.init.constant_(self.keypoint_center_conv.bias, bias_value)
torch.nn.init.constant_(self.heatmap_conv.bias, bias_value)

File: mmpose/models/heads/heatmap_heads/cid_head.py (hunk 1 above)
Comment (Ben-Louis) [reply mmpose/models/heads/heatmap_heads/cid_head.py:581]: For example, consider x. Its range in the input image space is [0, w-1], but in the heatmap space, it is [0, w-4] after being rescaled. The shift is (4-1)/2=1.5. 

However, I have found that using `preds.keypoints[..., :2] += 2` results in higher accuracy. I will fix this in a later commit. Do you have any idea about the reason?

File: mmpose/models/heads/heatmap_heads/cid_head.py (hunk 2 above)
Comment (Ben-Louis) [reply mmpose/models/heads/heatmap_heads/cid_head.py:334]: This initialization has not been added yet because the alignment of training accuracy has not been started. Thanks for your reminder

File: mmpose/models/heads/heatmap_heads/cid_head.py (hunk 1 above)
Comment (jin-s13) [reply mmpose/models/heads/heatmap_heads/cid_head.py]: 0.5 * 4 (down-sampling scale) = 2?

File: mmpose/models/losses/classification_loss.py [hunk 3]
```diff
@@ -153,3 +153,41 @@ def forward(self, pred_simcc, gt_simcc, target_weight):
                 self.criterion(coord_y_pred, coord_y_gt).mul(weight).sum())
 
         return loss / num_joints
+
+
+@MODELS.register_module()
+class InfoNCELoss(nn.Module):
+    """InfoNCE loss for training a discriminative representation space with a
+    contrastive manner.
```
Comment (ly015) [inline mmpose/models/losses/classification_loss.py:161]: Could you please add the paper reference for InforNCELoss and FocalHeatmapLoss?

Comment (Ben-Louis) [reply mmpose/models/losses/classification_loss.py:161]: Sure. The references have been added

**Review by ly015 — APPROVED** [review]
